### DISCUSSION TITLE
Contextvars are not reset between requests if the request contains multipart form data

### DISCUSSION THREAD

**Discussion by simon-sk** [Kludex/uvicorn#2152] [Potential Issue] [answered]
Hi uvicorn team,

we are experiencing an issue with uvicorn where `contextvars` set in a middleware are leaking between requests if the request contains multi-part form data and the client session is reused.

I constructed a minimal example which sets a context var in a middleware using Starlette and resets it after the app has been called. It also raises an exception if the var is already set:
```python
import contextvars

from starlette.applications import Starlette
from starlette.middleware import Middleware
from starlette.responses import JSONResponse
from starlette.routing import Route

var = contextvars.ContextVar("var", default=None)


class ASGIMiddleware:
    def __init__(self, app):
        self.app = app

    async def __call__(self, scope, receive, send):
        if scope["type"] != "http":
            return await self.app(scope, receive, send)

        if var.get() is not None:
            raise ValueError(f"var is already set {var.get}")

        token = var.set("TEST")

        await self.app(scope, receive, send)

        var.reset(token)


async def route(request):
    return JSONResponse({"hello": "world"})


app = Starlette(
    middleware=[Middleware(ASGIMiddleware)],
    routes=[Route('/', route, methods=["POST"])]
)


if __name__ == "__main__":
    import uvicorn

    uvicorn.run(
        "app.main_starlette:app",
        host="0.0.0.0",
        port=8001,
    )
```

Then, I execute a second script which calls this endpoint using a `httpx` client and includes a files object:
```python
import httpx

if __name__ == "__main__":
    url = "http://localhost:8001/"
    files = {'file': open('./test.yaml', 'rb')}

    with httpx.Client() as client:
        for i in range(2):
            r = client.post(url, timeout=None, files=files)
            assert r.status_code == 200
```

This results in a server error due to the fact that the variable is already set (sometimes it is necessary to run the client script multiple times as the issue does not always occur).

If I execute the same script without multipart data, it works without issues:
```python
...
r = client.post(url, timeout=None)
...
```

We have also experienced this issue without sending multipart form data and just sending a large request body but it occurs less frequently.

We experienced this issue with Python 3.9 and 3.10, Starlette v0.31.1, and uvicorn v0.23.2.

**Comment by Kludex** [comment]
Related: https://github.com/encode/starlette/discussions/2312

**Reply by Kludex** [reply, marked as answer]
I've promoted the issue on Starlette to an issue.

**Reply by simon-sk** [reply]
Perfect, thanks @Kludex! Since opening the discussion in the Starlette repo we have also found out that it occurs fairly consistently for large request bodies which is missing in the description of the issue there:
```
We have also experienced this issue without sending multipart form data and just sending a large request body but it occurs less frequently.
```

**Reply by dolfinus** [reply]
Maybe related: https://github.com/encode/uvicorn/discussions/2044

**Reply by simon-sk** [reply]
@dolfinus Yes, this looks exactly like the issue we experience as it also only occurs for really large payloads
